### ISSUE TITLE
automate and scrap rotating domains of the disposable providers

### ISSUE THREAD

**Issue by martenson** [disposable-email-domains/disposable-email-domains#450]
Some disposable services have fairly stable urls with lists -- hence we could implement a CI check that automatically merges domains in that list with our blocklist or opens PRs. Places that do not have such a resource can still be scrapped in CI with a headless browser or such.

**Comment by haumacher** [comment]
Is there already a list of disposable e-mail provider web pages that could serve as input for such automation?

**Comment by martenson** [comment]
@haumacher I'd start with some of the bigger providers like 

`https://10minutemail.com/` 
`https://m.kuku.lu/`
`https://temp-mail.org/`

**Comment by maciejstromich** [comment]
another one is https://www.disposablemail.com/

**Comment by haumacher** [comment]
I did some research - analyzing the commit log of `disposable_email_blocklist.conf` and extracting all URLs of fake mail service providers mentioned in the commit message. This gave the following fake mail services:

https://10minemail.com
https://10minutemail.com
https://10minutemail.net
https://10minutemail.org
https://10minutesmail.net
https://1secmail.com
https://2chmail.net
https://developermail.com
https://disposemymail.com
https://docs.webhook.site
https://dropmail.me
https://email-fake.com
https://emailfake.com
https://emailtemporal.org
https://etempmail.com
https://fakermail.com
https://linshiyou.com
https://linshiyouxiang.net
https://m.kuku.lu
https://mail.gw
https://mail.td, https://mail.cx
https://mail.tm
https://mohmal.com
https://muellmail.com
https://nospam.today
https://receivemail.org
https://ruu.kr
https://snapmail.cc
https://sute.jp
https://temp-mail.org
https://temp-mail.us
https://temp-mail.winsub.kr
https://tempmail.click
https://tempmail.lol
https://tempmail.plus
https://tempmailer.net
https://tempmailo.com
https://temporary-mail.net
https://tempr.email
https://trashmail.com
https://trashmailgenerator.de
https://www.disposablemail.com
https://www.emailnator.com
https://www.emailondeck.com
https://www.fakemail.net
https://www.linshi-email.com
https://www.minuteinbox.com
https://www.onetime-mail.com
https://www.trash-mail.com
https://youxiang.dev
https://ワンタイムメール.総合サービス.com

**Comment by icyavocado** [comment]
Should the script open a web headless browser and use a regex to search for the new email list. Adding the new domain, and then commit the new diff as a new PR?

**Comment by haumacher** [comment]
@icyavocado I think, this would be a huge effort and only possible for dumb fake-mail providers. "Professional" ones have strong protection against automated querying e.g. CAPTCHAs or even headless browser detection.

Here are some more:

https://www.spamgourmet.com
https://www.fakemailgenerator.com
https://www.easytrashmail.com
https://www.txen.de
https://www.throwawaymail.com
https://hi2.in
https://temp-mail.io
http://ese.kr
https://ulm-dsl.de/

**Comment by icyavocado** [comment]
Agree, from what I'm seeing, this may be a case of high effort, low reward scenario without a way to bypass bot checking. Unless we can show with certainty that requests from Github to these domains won't get blocked, finding an easy way to carry out this change doesn't seem likely.

**Comment by martenson** [comment]
I agree that some are well-enough protected against headless browser checking for domains. However, I believe there still are some low hanging fruits to be found. E.g. `https://www.fakemailgenerator.com/` just directly gives you the dropdown with domain selection. Writing a selenium case that loads this and then checks the list against blocklist opening PR if there is a diff is a nice step.

**Comment by icyavocado** [comment]
Here is my proposal for the script using puppeteer. Here is the proposed change: https://github.com/icyavocado/disposable-email-domains/commit/384f7a8380fdeee17a903fa873b97a2bf2d2d0aa

TLDR: this script reads from the `disposable_providers.txt` file and attempts to identify disposable email addresses. It ensures that any domains already on our blocklist or allowlist are excluded from this process.

Example output:
```bash
> disposable-email-domains (master) node fetch_email.js
No email domain found on https://10minemail.com
No email domain found on https://10minutemail.com
Domains already exist https://10minutemail.net
Found [ '10minutemail.org' ] domains on https://10minutemail.org
No email domain found on https://10minutesmail.net
No email domain found on https://1secmail.com
Found [ 'e.2chmail.net' ] domains on https://2chmail.net
No email domain found on https://developermail.com
```

This is just the first step of the task. As we continue our discussion about how to implement this correctly, I'll be working on the automation aspect.

**Comment by SteeBono** [comment]
I ran some tests and noticed that most domains associated with temporary emails have landing pages explaining that the domain is used for temporary email services. However, other domains lack a main page and return a 404 error instead.

To automatically exclude such domains, we could automate the process by performing a GET request to the domain and using a regex to check the HTML content for keywords like “temp,” “temporary,” “disposable,” etc.

We probably won’t exclude all temporary emails, but we could filter out a significant portion of them. What do you think?

**Comment by martenson** [comment]
@SteeBono the fact a domain has no landing on http has nothing to do with it being used for a disposable email service

**Comment by eventcountdown123-rgb** [comment]
Hey, I ran into this same issue with spam signups recently. The problem is that simple Regex validation passes most disposable domains because they have valid MX records.

I ended up building a small API to check domains against a live blocklist of ~60k burner providers. It filters out the subtle ones like `xy7-temp.com` that usually slip through.

It solved the issue for me, so I opened up the API for others to use. You can check it out here if you want to test it: [https://vishodi.com/](https://vishodi.com/disposable-email-detection) (There's a free tier for devs).

Hope it helps!

**Comment by eason4kim-rocket** [comment]
Scraping provider landing pages with headless browsers is always a perpetual game of whack-a-mole because modern disposable services rotate ephemeral subdomains and register dozens of fresh TLDs daily.

A much more scalable and maintainable detection approach is **MX Cluster & ASN Fingerprinting**:

### 1. MX Host Clustering (Catching 100s of domains from 1 record)
Disposable providers almost never run separate mail exchange infrastructure for every domain they register. They point dozens or hundreds of ephemeral burner domains to a small cluster of shared MX hosts or ASNs (e.g. `mail.guerrillamail.com`, `inbound.tempmail.ninja`, `smtp.yopmail.com`).

Instead of scraping HTML, resolving the MX record of new candidate domains and matching against known **MX root fingerprints** instantly flags entire domain families without needing manual per-provider scraping rules.

### 2. Multi-Signal Verification to Prevent False Positives
Pure scraping risks accidentally flagging legitimate privacy relays or university forwarding domains:
- **Signal 1: DNS MX & NULL-MX (RFC 7505) verification** (many scraped domains don't even have active mail routing).
- **Signal 2: MX Host ASN clustering** to detect shared disposable infrastructure.
- **Signal 3: Role-based account tagging** (`admin@`, `support@`, `billing@`).

In our email validation pipelines (e.g. what we built in LookTwice), combining DNS MX cluster heuristics with static blocklists reduced false positives significantly while catching newly minted burner domains on day 0 before they ever appear in static lists.
